### PR TITLE
Add "detachedAt" property to deleted repositories

### PR DESCRIPTION
**Pull request by DorianSyruss**
This PR solves issue #53. Upon repo deletion, publishes the new version of catalog with "deprecatedAt" attribute added for the deleted repository. The publishing should only happen if the deleted repo was already published.

### CODE REVIEW

File: server/course/course.controller.js [hunk 1]
```diff
@@ -54,6 +54,11 @@ function patch({ body, course, user }, res) {
 
 function remove({ course, user }, res) {
   return course.destroy({ context: { userId: user.id } })
+    .then(course => {
+      const deprecatedAt = course.dataValues.deleted_at;
```
Comment (DorianSyruss) [inline server/course/course.controller.js]: If anyone knows why is only snake case "deleted_at" field populated and the change only get's reflected in dataValues (on returned instance meta deleted_at doesn't get populated, nor aliased), help would be appreciated. @vladimyr

File: server/shared/publishing/helpers.js [hunk 2]
```diff
@@ -56,6 +56,18 @@ function updateRepositoryCatalog(repository, publishedAt) {
   });
 }
 
+function deprecateRepository(repository, deprecatedAt) {
```
Comment (DorianSyruss) [inline server/shared/publishing/helpers.js]: I've separated concerns of repo deprecation and repo update with this method, even though we could probably refactor "updateRepositoryCatalog" method to deprecate repos under certain conditions but I'm not sure we should.

File: server/course/course.controller.js [hunk 3]
```diff
@@ -54,6 +54,11 @@ function patch({ body, course, user }, res) {
...
+      if (!deprecatedAt) return;
+      publishingService.deprecateRepository(course.dataValues, deprecatedAt);
```
Comment (kronicker) [inline server/course/course.controller.js]: Move this call to `afterDestroy` hook.

File: server/course/course.controller.js (hunk 1 above)
Comment (DorianSyruss) [reply server/course/course.controller.js]: This is resolved, by passing ```returning: true``` in options instance gets returned with exact data.

File: server/shared/publishing/helpers.js [hunk 4]
```diff
@@ -41,11 +41,21 @@ function publishActivity(activity) {
   });
 }
 
-function updateRepositoryCatalog(repository, publishedAt) {
+function getRepositoryCatalog() {
   return storage.getFile('repository/index.json').then(buffer => {
-    let catalog = (buffer && JSON.parse(buffer.toString('utf8'))) || [];
+    return (buffer && JSON.parse(buffer.toString('utf8'))) || [];
```
Comment (zcuric) [inline server/shared/publishing/helpers.js]: Can you make this return more readable? Split it, maybe.

File: server/script/detachDeletedRepos.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+const { Course } = require('../shared/database');
+const each = require('lodash/each');
+const omit = require('lodash/omit');
+const find = require('lodash/find');
+const storage = require('../shared/storage');
+const { getRepositoryAttrs, getRepositoryCatalog } = require('../shared/publishing/helpers');
```
Comment (abasic) [inline server/script/detachDeletedRepos.js]: Multiple destructures should be on top.

**Comment by underscope** [comment]
@DorianSyruss Please merge develop

File: server/course/course.model.js [hunk 6]
```diff
@@ -67,6 +68,15 @@ class Course extends Model {
     });
   }
 
+  static hooks() {
+    return {
+      afterDestroy(course) {
+        return Course.findByPk(course.id, { paranoid: false })
```
[the remark this answers is not in the document]
Comment (DorianSyruss) [reply server/course/course.model.js]: Exactly, could have done instance.reload but It's the same I guess, and already saw this pattern on tailor.

**Pull request merged**